modules/Bio/EnsEMBL/Mapper.pm

```diff
@@ -282,50 +284,55 @@ sub map_coordinates {
     throw("Type $type is neither to or from coordinate systems");
   }
 
+  my @result;
+  my @paired_result;
+
   if ( !defined $hash->{ uc($id) } ) {
     # one big gap!
     my $gap = Bio::EnsEMBL::Mapper::Gap->new( $start, $end );
-    return $gap;
+    if ($include_original_region) {
+      push @paired_result, { 'original' => $gap, 'mapped' => $gap };
+      return @paired_result;
+    } else {
+      push @result, $gap;
+      return @result;
+    }
   }
 
   my $last_used_pair;
-  my @result;
-  my @paired_result;
 
-  my ( $start_idx, $end_idx, $mid_idx, $pair, $self_coord );
   my $lr = $hash->{ uc($id) };
+  
+  # create set of intervals to be checked for overlap
+  my $from_intervals;
 
-  $start_idx = 0;
-  $end_idx   = $#{$lr};
+  foreach my $i (@{$lr}) {
+    my $start = $i->{$from}{start}<=$i->{$from}{end}?$i->{$from}{start}:$i->{$from}{end};
```

Comment:
This isn't the clearest. Could this and the line below get split into an single if. The `$start` and `$end` can get created as `$i->{$from}{start}` and `$i->{$from}{end}` respectively then modified if they fail the if condition. This would also reduce the checks

Reply:
Thanks Mark, this is addressed now.

Reply:
That look much better, thank you

Reply:
The points against using an off the shelf implementation:
- minimise reliance on external libraries (i.e. Set::IntervalTree)
- Tree::Interval only handles the non-overlapping intervals case
- ability to provide optimize (i.e. XS) code, when possible
- ability to provide different suitable implementations (i.e. mutable/immutable) depending on the use case